File: DCTAuth/_DCTAuthMultipartData.m
//
//  _DCTAuthMultipartData.m
//  DCTAuth
//
//  Created by Daniel Tull on 02/09/2012.
//  Copyright (c) 2012 Daniel Tull. All rights reserved.
//

#import "_DCTAuthMultipartData.h"

@implementation _DCTAuthMultipartData

- (NSData *)dataWithBoundary:(NSString *)boundary {
	NSMutableData *body = [NSMutableData new];
	[body appendData:[[NSString stringWithFormat:@"--%@\r\n", boundary] dataUsingEncoding:NSUTF8StringEncoding]];
	[body appendData:[[NSString stringWithFormat:@"Content-Disposition: form-data; name=\"%@\"\r\n", self.name] dataUsingEncoding:NSUTF8StringEncoding]];
	[body appendData:[[NSString stringWithFormat:@"Content-Type: %@\r\n", self.type] dataUsingEncoding:NSUTF8StringEncoding]];
	[body appendData:[@"\r\n" dataUsingEncoding:NSUTF8StringEncoding]];
	[body appendData:self.data];
	[body appendData:[@"\r\n" dataUsingEncoding:NSUTF8StringEncoding]];
	return [body copy];
}

@end

File: DCTAuth/DCTAuthMultipartData.swift

import Foundation

class DCTAuthMultipartData: NSObject {

	var data: NSData?
	var name: String?
	var type: String?

	func dataWithBoundary(boundary: String) -> NSData? {

		let part1 = "--\(boundary)\r\n".dataUsingEncoding(NSUTF8StringEncoding)
		let part2 = "Content-Disposition: form-data; name=\"\(name)\"\r\n".dataUsingEncoding(NSUTF8StringEncoding)
		let part3 = "Content-Type: \(type)\r\n".dataUsingEncoding(NSUTF8StringEncoding)
		let part4 = "\r\n".dataUsingEncoding(NSUTF8StringEncoding)
		let part5 = "\r\n".dataUsingEncoding(NSUTF8StringEncoding)

		if let data = data, part1 = part1, part2 = part2, part3 = part3, part4 = part4, part5 = part5 {


			let parts = [part1, part2, part3, part4, data, part5]
			let body = parts.reduce(NSMutableData()) { (data, part) in
				data.appendData(part)
				return data
			}

			return body.copy() as? NSData

		}

		return nil
	}
}
